README.md

```diff
@@ -1,16 +1,611 @@
+
+### 3.1.4 Nonce
+
+The `nnc` field MUS contain a unique nonce. This field makes each invocation unique.
+
+### 3.1.5 Extended Fields
+
+The OPTIONAL `ext` field MAY contain arbitrary data. If not present, the `ext` field MUST me interpreted as `null`, including for [signature](#315-signature).
+
+### 3.1.6 Signature
+
+The `sig` field MUST contain the signature of the other fields. To construct the payload, the other fields MUST first be canonicalized as `dag-cbor`.
+
+If serialized as JSON, the `sig` field MUST be serialized as [unpadded base64url](https://datatracker.ietf.org/doc/html/rfc4648#section-5).
+
+## 3.2 IPLD Schema
+
+``` ipldsch
+type Invocation struct {
+  inv &UCAN  -- The UCAN providing authority
+  v   SemVer -- Version
+  run Scope  -- Which actions to invoke
+  nnc String -- Nonce
+  ext Any    -- Extended fields
+  sig Bytes  -- Signature
+}
+
+type Scope enum {
+  | All ("*")
+  | Action
+}
+
+type Action enum {
+  using  URI 
+  do     Ability
+  inputs Any
+} 
+```
+
+## 3.3 JSON Examples
+
+### 3.3.1 Run All
+
+ ``` js
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+ ``` js
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "run": "*", // Exmplicitely "run all"
+  "ext": null,
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+### 3.3.2 Pipelines
+
+The following examples both express the following dataflow graph:
+
+```
+              ┌────────────────────────────┐
+              │                            │
+              │ dns://example.com?TYPE=TXT │
+              │        crud/update         │
+              │                            │
+              └───────┬──────────┬─────────┘
+                      │          │
+                      │          │
+┌─────────────────────▼───┐  ┌───▼────────────────────────┐
+│                         │  │                            │
+│ mailto:alice@exaple.com │  │ mailto://alice@example.com │
+│         msg/send        │  │          msg/send          │
+│     bob@example.com     │  │      carol@exmaple.com     │
+│                         │  │                            │
+└─────────────────────┬───┘  └───┬────────────────────────┘
+                      │          │
+                      │          │
+             ┌────────▼──────────▼────────┐
+             │                            │
+             │ https://example.com/events │
+             │         crud/create        │
+             │                            │
+             └────────────────────────────┘
+```
+
+#### 3.3.2.1 Batched 
+
+ ``` json
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "run": {
```

Comment:
It is starting to look like GraphQL 😍 Some of my early [ucanto sketches](https://github.com/web3-storage/ucanto/blob/eef71dd602f59e1bab55e2661bc48ecd6177cee4/packages/interface/src/query.ts#L199-L216) were heading similar direction, but don’t end up going that direction to cut the scope.

This makes me wonder whether it would make more sense to embrace something like [edn query used by datomic](https://docs.datomic.com/cloud/whatis/edn.html#query-edn-example) or an [EQL](https://edn-query-language.org/eql/1.0.0/what-is-eql.html) as both have to deal with defining which I’m sure overlaps in various ways with dialog

Reply:
I really like EDN! We're not going to ship EDN as a first syntax for Dialog because the overall feedback we've received is that it's too unfamiliar for normies.

Reply:
As far as I know, EDN doesn't let you easily reference external facts, so we'd need to build our own calling convention on top of EDN. You're right that we're going to have the equivalent functionality in Dialog, but probably without EDN as a synatx to get started.

EDN also supports things like recursive queries, which I don't think we need?

Its userbase is also tiny; I don't know if it helps us on a familiarity font.

HOWEVER, I wonder if we can "just" use paths like IPNS does. `/action-name` as a relative link, and `/QmFooBar/action-name` for external ones. Strictly speaking, I think you'd have to include the `run` segment as well (`/QmFooBar/run/action-name`), which I like less if we later change the calling convention. It's also a path to the output, not the actual data, so would we need to signal that we're not looking for the data at that path, but the output associated with it? We can't have someone reference the action _directly_ because each run is unique, even if called with the same args (e.g. HTTP POST is not idempotent), so now we'd need to put a nonce in there (which makes life a bit more difficult in IPVM and Bacalhau as well, since we want to cache outputs).

Reply:
I don’t mean adopting EDN as is necessarily, but perhaps JSON-ified version could work here something along the lines of https://www.instantdb.com/essays/datalogjs 

I’m just afraid that this may grow into another almost query language at which point something not as familiar for JS crowd still may be better than something entirely new.

Reply:
> perhaps JSON-ified versio

Ah, sorry, I should have been clearer: I meant the JSONified version in my earlier messages.

> I’m just afraid that this may grow into another almost query language

Fair! I was trying to keep it to a selector language rather than a query language to reduce scope and keep it easy to learn and implement. But perhaps I'm going about it all wrong!

@QuinnWilton, I summon thee to the thread 🪄 You've looked at a few query languages recently, any thoughts?

Reply:
CORRECTION: I summon thee _during working hours_

Reply:
Also FWIW the answer for really complex data queries and transformation may be something like "use IPVM, or avoid the pipeline because you need to do something really complex". Even in our Dialog work, we're exploring user defined functions that call out to Wasm.

Reply:
> Also FWIW the answer for really complex data queries and transformation may be something like "use IPVM, or avoid the pipeline because you need to do something really complex".

This is my intuition too! Adding a custom query language to the base spec increases implementation burden considerably, and also increases the threat surface for authorization bypasses and denial of service attacks. Even something as simple as regular expression support opens the door for [ReDoS](https://en.wikipedia.org/wiki/ReDoS) attacks if you aren't careful.

Since this spec isn't focused on scheduling at all, from what I can tell, I suspect that it's the wrong place to consider problems like gas and timeouts, and so I'd argue that logic of any complexity belongs at another layer, like IPVM, that's more focused on capturing those sorts of runtime characteristics.

That being said, simple standards, like JSON Pointer or JSON Path (which isn't a standard, but is common and with decent library support), offer an easy way to get some lightweight support for queries and transclusion, but without adding much implementation burden, or exposing too much additional risk over a more static system.